jax/api.py

```diff
@@ -877,6 +877,12 @@ def vmap(fun: Callable, in_axes=0, out_axes=0) -> Callable:
     msg = ("vmap out_axes must be an int, None, or (nested) container with "
            "those types as leaves, but got {}.")
     raise TypeError(msg.format(out_axes))
+  if not all(l >= 0 for l in tree_leaves(in_axes) if isinstance(l, int)):
```

Comment:
How about checking it as `if any(l < 0 for l in tree_leaves(in_axes) if l is not None)`? It's more constrained that way, and potentially also faster (though I'm sure there are much bigger overheads here than `isinstance`). Also it's nice to allow short-circuiting behavior by checking `any` rather than `not all`.

Reply:
Actually, we don't even need the `is not None` filter: `None`s are treated as empty pytrees, so they're already filtered out of the leaves:

```python
In [6]: tree_leaves([None, None])
Out[6]: []
```

So how about `if any(l < 0 for l in tree_leaves(in_axes))`?

Reply:
Or maybe we need to filter `if l is not batching.last`...

Reply:
I can take a stab at this tweak!